### PR TITLE
rewrite umx loader to use the embedded offset, size and object type

### PR DESCRIPTION
**Pull request by sezero**
... instead of linear search.

Similar versions are in libmikmod and libmodplug. For the
upkg parser code, see:  https://github.com/sezero/umr.git

### CODE REVIEW

File: src/loaders/umx_load.c [hunk 1]
```diff
@@ -43,97 +42,374 @@ const struct format_loader libxmp_loader_umx = {
...
+	"IT", "S3M", "XM", "MOD",
+	"WAV", "MP2", NULL
+};
+
+/* decode an FCompactIndex.
+ * original documentation by Tim Sweeney was at
+ * http://unreal.epicgames.com/Packages.htm
+ * also see Unreal Wiki:
+ * http://wiki.beyondunreal.com/Legacy:Package_File_Format/Data_Details
+ */
```
Comment (AliceLR) [inline src/loaders/umx_load.c:100]: Not a real review yet (sorry, it is late, will do tomorrow) but I checked docs/formats/ and currently there is no UMX doc. Both of these links seem dead, but would it be worth pulling one of these from Wayback (or somewhere else)?

Comment (sezero) [reply src/loaders/umx_load.c:100]: Wayback links: 
https://web.archive.org/web/20080122034944/http://unreal.epicgames.com/Packages.htm
https://web.archive.org/web/20201023045227/http://wiki.beyondunreal.com/Legacy:Package_File_Format
https://web.archive.org/web/20201023062426/http://wiki.beyondunreal.com/Legacy:Package_File_Format/Data_Details

Don't know whether it's worth adding something under docs/

Comment (sezero) [reply src/loaders/umx_load.c:100]: OK, I copied the attached txt from beyondunreal wiki. 
Would it be appropriate to add under docs/?
[upkg.txt](https://github.com/libxmp/libxmp/files/5911580/upkg.txt)

Comment (sezero) [reply src/loaders/umx_load.c:100]: Well, I went ahead and added the documentation under docs/formats/ with the last commit.

File: src/loaders/umx_load.c [hunk 2]
```diff
@@ -43,97 +42,374 @@ const struct format_loader libxmp_loader_umx = {
...
-	uint8 buf[TEST_SIZE], *b = buf;
-	uint32 id;
 
-	LOAD_INIT();
+	/* byte swap the header - all members are 32 bit LE values */
+	p = (unsigned char *) header;
+	swp = (uint32 *) header;
+	for (i = 0; i < UPKG_HDR_SIZE/4; i++, p += 4) {
+		swp[i] = p[0] | (p[1] << 8) | (p[2] << 16) | (p[3] << 24);
+	}
```
Comment (AliceLR) [inline src/loaders/umx_load.c:301]: This seems questionable—`header` here is a stack allocated char array passed from `process_upkg` and I suspect it is not guaranteed to be DWORD aligned. Would it be less confusing (and safer) to just read these values directly into a `upkg_hdr` struct with `hio_read32l`?

**Comment by AliceLR** [comment]
I set up Unreal Gold and Unreal Tournament to test their music with this and predictably everything still works. Aside from the questionable byteswapping in `probe_header` I think this patch looks OK.

File: src/loaders/umx_load.c (hunk 2 above)
Comment (sezero) [reply src/loaders/umx_load.c]: Done with the new commit: please review.

**Review by AliceLR — APPROVED** [review]
This looks good to me.

**Comment by sezero** [comment]
> This looks good to me.

Thanks. Will merge as soon as Travis completes.

**Pull request merged**